FF-A S-EL1 SPMC

jenswi-linaro opened the pull request:
QEMU v8 FF-A S-EL1 SPMC prototype can be fetched with:
repo init -u https://github.com/jenswi-linaro/manifest.git -m qemu_v8.xml -b ffa_sel1_spmc

jforissier commented on core/arch/arm/mm/sub.mk:
```diff
@@ -9,7 +9,11 @@ endif
 srcs-y += pgt_cache.c
 srcs-y += mobj.c
+ifeq ($(CFG_CORE_SPCI),y)
+srcs-y += mobj_ffa.c
```
Can we make this `srcs-$(CFG_CORE_FFA) += mobj_ffa.c` with no `ifeq`?
It is always a bit suspect when the name of the file does not reflect the config name.

jenswi-linaro replied:
Sorry, a leftover that got overwritten in later commits, I'll fix.

jforissier commented on core/arch/arm/include/sm/psci.h:
```diff
@@ -26,6 +27,8 @@
 #define PSCI_FN_STAT_COUNT		PSCI_FN(17)
 
+
+
```
Extra lines not needed

jenswi-linaro commented:
Did a force push this time since I think it's easier at this step. Addressed comments (+checkpatch) and fixed up commit messages. Added a shippable target in the last commit.

etienne-lms commented on core/arch/arm/include/ffa.h:
```diff
@@ -0,0 +1,85 @@
+#define FFA_VERSION_MINOR_SHIFT		0
+#define FFA_VERSION_MINOR_MASK		U(0xFFFF)
+#define MAKE_FFA_VERSION(major, minor)	((((major) & FFA_VERSION_MAJOR_MASK)  \
+						<< FFA_VERSION_MAJOR_SHIFT) | \
```
`<<` should be in the previous line.
Maybe:
```c
#define MAKE_FFA_VERSION(major, minor)	\
	((((major) & FFA_VERSION_MAJOR_MASK) << FFA_VERSION_MAJOR_SHIFT) | \
	 ((minor) & FFA_VERSION_MINOR_MASK))

etienne-lms commented on core/arch/arm/include/ffa.h:
```diff
@@ -0,0 +1,85 @@
+#define FFA_VERSION_MAJOR_MASK		U(0x7FFF)
+#define FFA_VERSION_MINOR		U(0)
+#define FFA_VERSION_MINOR_SHIFT		0
+#define FFA_VERSION_MINOR_MASK		U(0xFFFF)
```
GENMASK_32(15, 0)

jenswi-linaro replied:
These simple masks are in my opinion easier to read if written plainly as here instead of using `GENMASK_32()`. I'll change it if you think it's important.

etienne-lms replied:
no strong opinion on this. Keep 0xFFFF if you think it's easier to read and maintain.

etienne-lms commented on core/arch/arm/include/ffa.h:
```diff
@@ -0,0 +1,85 @@
+/* FFA_VERSION helpers */
+#define FFA_VERSION_MAJOR		U(1)
+#define FFA_VERSION_MAJOR_SHIFT		16
+#define FFA_VERSION_MAJOR_MASK		U(0x7FFF)
```
GENMASK_32(14, 0)

etienne-lms commented on core/arch/arm/include/ffa.h:
```diff
@@ -0,0 +1,85 @@
+
+/* FFA_MSG_SEND helpers */
+#define FFA_MSG_SEND_ATTRS_BLK_SHIFT	U(0)
+#define FFA_MSG_SEND_ATTRS_BLK_MASK	U(0x1)
```
prefer BIT(0) as this is a bit field mask

jenswi-linaro replied:
I'll remove these helpers, it turned out the FFA_MSG_SEND is not used in secure world.

etienne-lms commented on core/arch/arm/include/ffa.h:
```diff
@@ -0,0 +1,85 @@
+#define FFA_MSG_SEND_ATTRS_BLK		U(0)
+#define FFA_MSG_SEND_ATTRS_BLK_NOT	U(1)
+#define FFA_MSG_SEND_ATTRS(blk)	(((blk) & FFA_MSG_SEND_ATTRS_BLK_MASK)	\
+					 << FFA_MSG_SEND_ATTRS_BLK_SHIFT)
```
`<<` should be in the previous line.

etienne-lms commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,202 @@
+ */
+
+/*
+ * This file is exported by OP-TEE and is in kept in sync between secure
```
s/in kept in/kept in/

jenswi-linaro replied:
Thanks

etienne-lms commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,202 @@
+
+/*
+ * This file is exported by OP-TEE and is in kept in sync between secure
+ * world and normal world kernel driver. We're using ARM FF-A 1.0 EAC
```
I would remove "kernel driver" to be more generic, consider normal world bootloader for example.

jenswi-linaro replied:
Agree, I'll fix

etienne-lms commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,202 @@
+ *
+ * Return register usage:
+ * w3:    Error code, 0 on success
+ * w4:    Bit[1:0]:  Number of pages of shared memory to register with
```
s/pages/4kB pages/ ?
Or maybe explicit in header inline comments (lines 7-9) that `page` in the file implicitly refer to 4kByte pages.
Or maybe this is implicit...

jenswi-linaro replied:
I'll update, better to be clear.

etienne-lms commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,202 @@
+ * with a zero internal offset and normal cached memory attributes.
+ * Register usage:
+ * w3:    Service ID, OPTEE_FFA_YIELDING_CALL_WITH_ARG
+ * w4:    Shared memory handle, lower bits
```
32bit LSB?

jenswi-linaro replied:
I'll try to make it more clear.

etienne-lms commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,202 @@
+ *
+ * Call to register shared memory memory. The data is supplied in shared
+ * memory with a zero internal offset and normal cached memory attributes.
+ * The data is formatted as described in SPCI 1.0 Beta1 Table 137 "
```
suggest to move character `"` to the next line.

etienne-lms commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,202 @@
+ * w5:    Shared memory handle, higher bits
+ * w6-w7: Not used (MBZ)
+ *
+ * Call unregister shared memory register usage:
```
Call **to** unregister shared memory **. Register** usage:

etienne-lms commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,202 @@
+ *	w4:    Shared memory handle, lower bits
+ *	w5:    Shared memory handle, higher bits
+ *	       If the allocation failed both w4 and w5 are 0
+ *	w6:    Offset into shared memory pointing the table If internal
```
Add a dot `.` before `If internal...`.

etienne-lms commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,202 @@
+ * OPTEE_FFA_YIELDING_CALL_REGISTER_SHM and
+ * OPTEE_FFA_YIELDING_CALL_UNREGISTER_SHM
+ * FFA_BUSY:               Number of OP-TEE OS threads exceeded,
+ *                          try again later
```
indentation: remove a space char.

etienne-lms commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,202 @@
+ *
+ * Possible error codes for OPTEE_FFA_YIELDING_CALL_RESUME
+ * FFA_INVALID_PARAMETER:  Bad resume info
+ *
```
remove line

etienne-lms commented on core/include/optee_msg.h:
```diff
@@ -126,10 +129,26 @@ struct optee_msg_param_rmem {
+ * @internal_offs: Internal offset into the first page of shared memory
+ *		   reference
+ * @size:	   Size of the buffer
+ * @global_id:	   Global identifier of Shared memory
```
s/of Shared/of the shared/

etienne-lms commented on core/arch/arm/mm/mobj_ffa.c:
```diff
@@ -0,0 +1,562 @@
+TEE_Result mobj_ffa_add_pages_at(struct mobj_ffa *mf, unsigned int *idx,
+				 paddr_t pa, unsigned int num_pages)
+{
+	unsigned int n;
```
`= 0`

etienne-lms commented on core/arch/arm/mm/mobj_ffa.c:
```diff
@@ -0,0 +1,562 @@
+	 * unregistered.
+	 */
+	if (mf) {
+		DMSG("cookie %#"PRIx64" busy refc %u",
```
suggest to move the trace out of the spinlock xsave.

jenswi-linaro replied:
Why? What's the concern?

etienne-lms replied:
I though about evacuating trace with interrupt masked.
Ok as is.

etienne-lms commented on core/arch/arm/mm/mobj_ffa.c:
```diff
@@ -0,0 +1,562 @@
+	}
+	mf = find_in_list(&shm_inactive_head, cmp_cookie, cookie);
+	/*
+	 * If the mobj isn't found or if it already have been unregistered.
```
s/have/has/

etienne-lms commented on core/arch/arm/mm/mobj_ffa.c:
```diff
@@ -0,0 +1,562 @@
+	}
+
+	mf = find_in_list(&shm_inactive_head, cmp_cookie, cookie);
+	/* If the mobj isn't found */
```
could remove this inline comment

etienne-lms commented on core/arch/arm/mm/mobj_ffa.c:
```diff
@@ -0,0 +1,562 @@
+			DMSG("cookie %#"PRIx64" active: refc %d",
+			     cookie, refcount_val(&mf->mobj.refc));
+		} else {
+			EMSG("cookie %#"PRIx64" mismatching internal_offs got %#"PRIx16" expected %#x",
```
Move traces should be outside the spinlock xsave?

jenswi-linaro replied:
Why? What's the concern?

etienne-lms replied:
ok as is

etienne-lms commented on core/arch/arm/mm/mobj_ffa.c:
```diff
@@ -0,0 +1,562 @@
+		goto out;
+	}
+
+	EMSG("cookie %#"PRIx64, mf->cookie);
```
s/EMSG/DMSG/

etienne-lms commented on core/arch/arm/mm/mobj_ffa.c:
```diff
@@ -0,0 +1,562 @@
+			 TEE_MM_POOL_NO_FLAGS))
+		panic("Could not create shmem pool");
+
+	DMSG("Shared memory address range: %"PRIxVA", %"PRIxVA,
```
s/%/%#/g

etienne-lms commented on core/arch/arm/include/arm.h:
```diff
@@ -45,6 +45,7 @@
 #define MPIDR_CLUSTER_SHIFT	MPIDR_AFF1_SHIFT
 #define MPIDR_CLUSTER_MASK	MPIDR_AFF1_MASK
 
+#define MPIDR_AARCH32_AFF_MASK	0xffffff
```
add empty line below + GENMASK_32(31, 0)

jenswi-linaro replied:
I'll add that and compose this define from MPIDR_AFF0_MASK..MPIDR_AFF2_MASK instead.

etienne-lms commented on core/arch/arm/plat-vexpress/fvp_spmc_pm.c:
```diff
@@ -0,0 +1,82 @@
+		if (a1 == mpidr)
+			continue;
+
+		DMSG("PSCI_CPU_ON op on mpidr 0x%lx\n", a1);
```
remove terminal `\n`

jenswi-linaro commented:
Update

imre-kis-arm commented on core/arch/arm/include/ffa.h:
```diff
@@ -0,0 +1,77 @@
+#define FFA_BUSY		-4
+#define FFA_INTERRUPTED		-5
+#define FFA_DENIED		-6
+#define FFA_RETRY		-7
```
There's also an ABORTED (-8) error code in the specification.

jenswi-linaro replied:
Thanks, I'll add it.

MrVan commented:
@jenswi-linaro 
A general question, ffa_sel1_spmc with QEMU v8 FF-A, is your model have secure EL2 enabled? Which arm trusted firmware branch are you using?

jenswi-linaro commented:
> A general question, ffa_sel1_spmc with QEMU v8 FF-A, is your model have secure EL2 enabled? Which arm trusted firmware branch are you using?

@MrVan, S-EL2 is not enabled in this case. As far a I know QEMU doesn't support S-EL2 yet. I'm using a temporary test branch for T-FA. You can try it out with:
```
repo init -u https://github.com/jenswi-linaro/manifest.git -m qemu_v8.xml -b ffa_sel1_spmc
repo sync -j4
cd build
make toolchains
make all -j4
make run-only
```

jenswi-linaro commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,202 @@
+ * w6:    Offset into shared memory pointing to a struct optee_msg_arg
+ * w7:    Not used (MBZ)
+ *
+ * Call to register shared memory memory. The data is supplied in shared
```
I'll fix

jenswi-linaro commented on core/arch/arm/include/optee_ffa.h:
```diff
@@ -0,0 +1,201 @@
+ *
+ * Call to register shared memory memory. The data is supplied in shared
+ * memory with a zero internal offset and normal cached memory attributes.
+ * The data is formatted as described in SPCI 1.0 Beta1 Table 137
```
Thanks

wmnt commented on core/arch/arm/kernel/thread_spmc.c:
```diff
@@ -0,0 +1,1158 @@
+
+		fpi->id = my_sp_id;
+		fpi->execution_context = CFG_TEE_CORE_NB_CORE;
+		fpi->partition_properties = BIT(0) | BIT(1) | BIT(2);
```
BIT(2) is MBZ according to the EAC FFA spec for SPs

jenswi-linaro replied:
Thanks, I'll fix.

wmnt commented on core/arch/arm/kernel/thread_spmc.c:
```diff
@@ -0,0 +1,1158 @@
+		 FFA_PARAM_MBZ, FFA_PARAM_MBZ, FFA_PARAM_MBZ);
+}
+
+static void handle_partition_info_get(struct thread_smc_args *args)
```
According to the FFA spec 
“A NS-Endpoint can request information for all partitions in the system including the caller by specifying the Nil UUID.”
and
“If the Nil UUID is specified at the Non-secure virtual FF-A instance, the Hypervisor must provide information for partitions resident in both security states.”
Consequently, if queried by a NS endpoint, the hypervisor must in turn query the SPM for all SPs by invoking this function with the Nil UUID.
In the case of this PR the only SP is OPTEE itself, so the same data must be returned to the caller for a Nil UUID as if the requested UUID matched that of OPTEE - otherwise the hypervisor would be unable to enumerate all the partitions in the system.

jenswi-linaro replied:
Thanks, I'll fix.

jenswi-linaro commented:
I'm going to squash and rebase the commits soon if there's no objection.

jenswi-linaro commented:
Squashed and rebased

etienne-lms commented on core/arch/arm/kernel/thread_spmc.c:
```diff
@@ -0,0 +1,1169 @@
+ * owned by normal world.
+ *
+ * Note that we can't prevent normal world from updating the content of
+ * these buffers so we must always be careful when reading.  while we hold
```
"when reading.  while we hold"  Maybe few missing words. or "  ." to remove.

etienne-lms commented on core/arch/arm/kernel/thread_spmc.c:
```diff
@@ -0,0 +1,1169 @@
+	uint32_t ret_w1 = 0;
+	uint32_t ret_w2 = 0;
+	uint32_t ret_w3 = 0;
+	uint32_t ret_fid =  0;
```
dual space character.
Here and lines 771, 774, 778, 788 and 797.

etienne-lms commented on core/arch/arm/kernel/thread_spmc.c:
```diff
@@ -0,0 +1,1169 @@
+	uint32_t ret_w1 = 0;
+	uint32_t ret_w2 = FFA_INVALID_PARAMETERS;
+	uint32_t ret_w3 = 0;
+	uint32_t ret_fid =  FFA_ERROR;
```
dual space char

etienne-lms commented on core/arch/arm/kernel/thread_spmc.c:
```diff
@@ -0,0 +1,1169 @@
+		goto out;
+	}
+	if (rc > 0) {
+		ret_fid =  FFA_MEM_FRAG_RX;
```
dual space char

etienne-lms commented on core/arch/arm/kernel/thread_spmc.c:
```diff
@@ -0,0 +1,1169 @@
+		ret_w3 = rc;
+		reg_pair_from_64(global_handle, &ret_w2, &ret_w1);
+	}
+	ret_fid =  FFA_SUCCESS_32;
```
dual space char

MrVan commented:
> > A general question, ffa_sel1_spmc with QEMU v8 FF-A, is your model have secure EL2 enabled? Which arm trusted firmware branch are you using?
> 
> @MrVan, S-EL2 is not enabled in this case. As far a I know QEMU doesn't support S-EL2 yet. I'm using a temporary test branch for T-FA. You can try it out with:
> 
> ```
> repo init -u https://github.com/jenswi-linaro/manifest.git -m qemu_v8.xml -b ffa_sel1_spmc
> repo sync -j4
> cd build
> make toolchains
> make all -j4
> make run-only
> ```

@jenswi-linaro So this needs platform-firmware in S-EL0? The platform firmare relys on OP-TEE to be the SPMC, right? Has ATF already support SPMC without S-EL2 feature?

jenswi-linaro commented:
Fixed up the commits with the recent comments and applied the tag.
I think this is as good as it gets right now, so this should be ready for merge.

jenswi-linaro commented:
> @jenswi-linaro So this needs platform-firmware in S-EL0? The platform firmare relys on OP-TEE to be the SPMC, right? Has ATF already support SPMC without S-EL2 feature?

This is just to allow normal world to use the FF-A ABI instead. For the time being this is experimental. Once there's proper support in upstream TF-A and Linux kernel we'll remove the experimental status.

jenswi-linaro commented:
@jforissier, OK to merge?

jforissier commented:
@jenswi-linaro fine with me. Sorry for the delay in processing pull requests, it is due to vacations.

jenswi-linaro commented:
> @jenswi-linaro fine with me. Sorry for the delay in processing pull requests, it is due to vacations.

No worries, it seems to work rather well. Do let us know if you need a temporary stand-in.
